test/e2e_node/device_plugin.go

```diff
@@ -0,0 +1,268 @@
+// that requests resourceName and runs the specified command.
+func makeBusyboxPod(resourceName, cmd string) *v1.Pod {
+	podName := "device-plugin-test-" + string(uuid.NewUUID())
+	rl := v1.ResourceList{v1.ResourceName(resourceName): *resource.NewQuantity(1, resource.DecimalSI)}
+
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{Name: podName},
+		Spec: v1.PodSpec{
+			RestartPolicy: v1.RestartPolicyAlways,
+			Containers: []v1.Container{{
+				Image: busyboxImage,
+				Name:  podName,
+				// Runs the specified command in the test pod.
+				Command: []string{"sh", "-c", cmd},
+				Resources: v1.ResourceRequirements{
+					Limits:   rl,
+					Requests: rl,
+				},
+			}},
+		},
+	}
+}
+
+// parseLogFromNRuns returns restart count of the specified container
+// after it has been restarted at least restartCount times,
+// and the matching string for the specified regular expression parsed from the container logs.
+func parseLogFromNRuns(f *framework.Framework, podName string, contName string, restartCount int32, re string) (int32, string) {
+	var count int32
+	// Wait till pod has been restarted at least restartCount times.
+	Eventually(func() bool {
+		p, err := f.PodClient().Get(podName, metav1.GetOptions{})
+		if err != nil || len(p.Status.ContainerStatuses) < 1 {
+			return false
+		}
+		count = p.Status.ContainerStatuses[0].RestartCount
+		return count >= restartCount
+	}, 5*time.Minute, framework.Poll).Should(BeTrue())
+	logs, err := framework.GetPodLogs(f.ClientSet, f.Namespace.Name, podName, contName)
+	if err != nil {
+		framework.Failf("GetPodLogs for pod %q failed: %v", podName, err)
+	}
+	framework.Logf("got pod logs: %v", logs)
+	regex := regexp.MustCompile(re)
+	matches := regex.FindStringSubmatch(logs)
+	if len(matches) < 2 {
+		return count, ""
+	}
+	return count, matches[1]
+}
+
+// numberOfDevices returns the number of devices of resourceName advertised by a node
+func numberOfDevices(node *v1.Node, resourceName string) int64 {
+	val, ok := node.Status.Capacity[v1.ResourceName(resourceName)]
+	if !ok {
+		return 0
+	}
+
+	return val.Value()
+}
+
+// stubAllocFunc will pass to stub device plugin
+func stubAllocFunc(r *pluginapi.AllocateRequest, devs []*pluginapi.Device) (*pluginapi.AllocateResponse, error) {
+	var response pluginapi.AllocateResponse
+	for _, requestID := range r.DevicesIDs {
+		dev := &pluginapi.Device{}
+		exists := false
+		for _, availableDevice := range devs {
+			if requestID == availableDevice.ID {
+				exists = true
+				dev = availableDevice
+				break
+			}
+		}
+
+		if !exists {
+			return nil, fmt.Errorf("invalid allocation request with non-existing device: %s", requestID)
+		}
+
+		if dev.Health != pluginapi.Healthy {
+			return nil, fmt.Errorf("invalid allocation request with unhealthy device: %s", requestID)
+		}
+
+		// create fake device file
+		fpath := filepath.Join(os.TempDir(), dev.ID)
+
+		// clean first
+		os.RemoveAll(fpath)
+		f, err := os.Create(fpath)
+		defer f.Close()
```

Comment:
If we close the file when the stubAllocFunc returns, would the later mount still succeed?

Reply:
Yes, this close just close `fd`, will not remove the file.

Reply:
@RenaudWasTaken Thanks for your good catch, will update this tomorrow.